### ISSUE TITLE
`ttnn.logical_not` (float32): returns 0 for `-0.0` — the signed-zero fix from #43831 was applied to the six zero-comparison ops but not to `calculate_logical_not`

### ISSUE THREAD

**Issue by beyondpremiere** [tenstorrent/tt-metal#56520]
Verified against tt-metal `main` @ `70f840f46adb33ec21ec8893f48ee06692648f75`, Blackhole ttsim v1.10.7, SFPI 7.76.0.
Evidence, reproducer and raw rows: https://gist.github.com/beyondpremiere/8477c0cccb2a3faaa224608723936f12

---

**Component / Area:** SFPU numerics — `tt_metal/hw/ckernels/{blackhole,wormhole_b0}/metal/llk_api/llk_sfpu/ckernel_sfpu_logical_not.h`
**Issue Type:** Correctness

## Summary

`calculate_logical_not` decides "is zero" with a raw SFPU compare:

```cpp
v_if(v == 0) { r = 1; }        // ckernel_sfpu_logical_not.h:37 (Blackhole), :39 (Wormhole)
```

On the float layout that compare is sign-bit ordered, so `-0.0` does not match and
`logical_not(-0.0)` returns `0`. IEEE 754, torch (`torch.logical_not(torch.tensor(-0.0))` is `True`)
and the harness's own golden (`_logical_not`: `1.0 if x == 0 else 0.0`) all give `1`.

This is the same defect #43831 reported for `eqz / nez / ltz / gtz / lez / gez`. That issue was fixed
(#44085, #44244) by classifying on `|v|` instead of `v` in `ckernel_sfpu_comp.h`, and those six ops
now return the right answer for `-0.0` (verified below). `calculate_logical_not` lives in its own header
and was not part of that change, so it still carries the pre-fix behaviour. `test_vif_equiv_sweep.py`
pins signed-zero expectations for `sign`, `heaviside`, `hardshrink` and the `unary_*` compares, and
`test_eltwise_unary_sfpu_zero_comp.py` covers the six zero-comparison ops; neither covers `logical_not`.

## Observed

Float32 in and out, `dest_acc=Yes` (the one path on which `-0.0` reaches the LREG with its sign bit —
`test_vif_equiv_sweep.py:210-212` says the same; on the bf16 path the unpacker flushes it to `+0.0`
first, so `logical_not` is coincidentally right there).

| input | bits | `logical_not` kernel | TT golden | torch |
|---|---|---|---|---|
| `+0.0` | `0x00000000` | 1 | 1 | 1 |
| `-0.0` | `0x80000000` | **0** | 1 | 1 |
| `±1.0`, `±inf` | — | 0 | 0 | 0 |
| NaN (9 payloads incl. negative) | — | 0 | 0 | 0 |

Same run, same input tile, the six ops fixed by #43831: `eqz(-0.0) = 1`, `nez(-0.0) = 0`,
`ltz(-0.0) = 0`, `lez(-0.0) = 1`, `gez(-0.0) = 1` — all correct. `logical_not` is the odd one out.

## Mechanism

`v == 0` lowers to `SFPSETCC`, whose ISA page states that "software should ensure that
sign-magnitude integers and FP32 have had negative zero flushed to positive zero" before the compare
(tt-isa-documentation, `WormholeB0/TensixTile/TensixCoprocessor/SFPSETCC.md`, summary; the harness quotes
the same contract in `test_vif_equiv_sweep.py:187-188`). The kernel therefore has to normalise the sign
itself before testing for zero; the zero-comparison kernels do (`|v| == 0`,
`ckernel_sfpu_comp.h:67-75`), `logical_not` does not.

## Proposed fix

Mirror the #43831 fix for the float layout only, leaving the integer layouts bit-exact:

```cpp
vType r = 0;
if constexpr (std::is_same_v<vType, sfpi::vFloat>) {
    v_if(sfpi::abs(v) == 0.0f) { r = 1; }
    v_endif;
} else {
    v_if(v == 0) { r = 1; }
    v_endif;
}
```

**Validated on the simulator:** with that change, `logical_not(-0.0) = 1`, and every other input in the
table (both zeros, `±1`, `±inf`, nine NaN payloads) is unchanged. Both architectures carry the same
`v_if(v == 0)` line and need the same edit. Suggested regression: add `logical_not` to the ops in
`test_vif_equiv_sweep.py`'s `_SIGNED_ZERO_EXPECTED` with `(1.0, 1.0)`.

## Reachability

`ttnn.logical_not` → `UnaryOpType::LOGICAL_NOT_UNARY` (`unary.cpp:155`) →
`SFPU_OP_LOGICAL_NOT_INCLUDE` / `logical_not_unary_tile` (`unary_op_utils.cpp:51, :788`) →
`calculate_logical_not` (`hw/inc/api/compute/eltwise_unary/logical_not.h:49`). `-0.0` is an ordinary
float32 result (`0.0 * -1`, `-x` of a zero, underflow of a negative product), so a mask built with
`logical_not` on such a tensor silently drops those lanes.

## Reproduction

`sweep_unary.py` + `stimuli.py` (drop into `tt_metal/tt-llk/tests/python_tests/`):

```bash
SWEEP_OPS=LogicalNot SWEEP_STIM=specials SWEEP_OUT=out CHIP_ARCH=blackhole \
  TT_METAL_SIMULATOR=/path/to/libttsim_bh.so pytest sweep_unary.py --run-simulator --forked
```

`baseline_LogicalNot_specials.csv` / `fixed_LogicalNot_specials.csv` are the recorded outputs (input
bits, output bits, golden) before and after `patch_signed_zero_fix.py`.

## Scope

Measured on the Blackhole functional simulator, not silicon. The mechanism is the documented ISA
contract, not a simulator artefact, but the numbers should be confirmed on hardware.

**Comment by shitobu** [comment]
Hey team 👋

Following up on the signed-zero work we just wrapped up on `ttnn.where` (#56538), I'd love to take this one on and get it sorted out.

The root cause is straightforward: in `calculate_logical_not` (`ckernel_sfpu_logical_not.h` on both Wormhole B0 and Blackhole), the raw `v_if(v == 0)` check evaluates sign-magnitude bits directly on FP32, so `-0.0` (`0x80000000`) fails the condition and returns `0` instead of `1.0`.

My plan is to:
1. Mirror the #43831 signed-zero normalization specifically for the float layout:
   ```cpp
   if constexpr (std::is_same_v<vType, sfpi::vFloat>) {
       v_if(sfpi::abs(v) == 0.0f) { r = 1; }
       v_endif;
   } else {
       v_if(v == 0) { r = 1; }
       v_endif;
   }
   ```
   leaving integer paths bit-exact.
2. Apply this cleanly across both Wormhole B0 and Blackhole LLK headers.
3. Add regression tests covering `-0.0` and `+0.0` in FP32 and BF16.

Could you please assign this issue to me? Ready to put up a clean PR whenever you're good.
